clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/gitRepo/GitJobExecutor.java

```diff
@@ -0,0 +1,294 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.clouddriver.artifacts.gitRepo;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.netflix.spinnaker.clouddriver.jobs.JobExecutor;
+import com.netflix.spinnaker.clouddriver.jobs.JobRequest;
+import com.netflix.spinnaker.clouddriver.jobs.JobResult;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.Charset;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FileUtils;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.util.StringUtils;
+
+@Slf4j
+public class GitJobExecutor {
+
+  private static final String SSH_KEY_PWD_ENV_VAR = "SSH_KEY_PWD";
+  private static Path genericAskPassBinary;
+
+  @Getter private final GitRepoArtifactAccount account;
+  private final JobExecutor jobExecutor;
+  private final String gitExecutable;
+  private final AuthType authType;
+  private final Path askPassBinary;
+
+  private enum AuthType {
+    USER_PASS,
+    TOKEN,
+    SSH,
+    NONE
+  }
+
+  public GitJobExecutor(
+      GitRepoArtifactAccount account, JobExecutor jobExecutor, String gitExecutable)
+      throws IOException {
+    this.account = account;
+    this.jobExecutor = jobExecutor;
+    this.gitExecutable = gitExecutable;
+    if (!StringUtils.isEmpty(account.getUsername())
+        && !StringUtils.isEmpty(account.getPassword())) {
+      authType = AuthType.USER_PASS;
+    } else if (!StringUtils.isEmpty(account.getToken())) {
+      authType = AuthType.TOKEN;
+    } else if (!StringUtils.isEmpty(account.getSshPrivateKeyFilePath())) {
+      authType = AuthType.SSH;
+    } else {
+      authType = AuthType.NONE;
+    }
+    askPassBinary = initAskPass();
+  }
+
+  public void clone(String repoUrl, String branch, Path destination) throws IOException {
+    if (!isValidReference(repoUrl)) {
+      throw new IllegalArgumentException(
+          "Git reference \""
+              + repoUrl
+              + "\" is invalid for credentials with auth type "
+              + authType);
+    }
+    FileUtils.deleteDirectory(destination.toFile());
+    FileUtils.forceMkdir(destination.toFile());
+    log.info("Cloning git/repo {} into {}", repoUrl, destination.toString());
+
+    String cloneCommand =
```

Comment:
It's not really fair of me to mention this since you've done this gigantic pile of very helpful work....Did you consider at all the additional juggling it would take to only clone the one subdirectory of a repo?  https://stackoverflow.com/questions/600079/how-do-i-clone-a-subdirectory-only-of-a-git-repository/52269934#52269934 has a massive essay about it, that I think boils down to:
```
    $ git clone --depth 1 --branch <specified branch> --filter=blob:none
    $ git checkout <specified branch> -- <specified path>/
```

Reply:
I did not consider that and just went with regular shallow clone, I'll see how (if?) that works

Reply:
Thank you!

Reply:
Just to add - the original implementation supported fetching (maybe it was just returning) a specific path via the `artifact.metadata.subPath` property.

Reply:
I saw that, but it was only for creating the `tgz` archive, not for cloning the repo. This implementation supports the same in the `git archive` command.

Reply:
@dbyron0 seems that this partial clone feature is not only dependent on the git client version (and `Dockerfile.ubuntu` doesn't have the min git version needed), but also on the git server supporting it properly. For instance, [Bitbucket doesn't support partial clones](https://jira.atlassian.com/browse/BSERV-11639?error=login_required&error_description=Login+required&state=54002a14-6fe4-4453-b0b5-90147953b3c1). 

So it would be better to work just with shallow clones now to ensure compatibility, and probably add a caching strategy if this is still a concern, so once the repo is cached we do a `git pull` every time it's needed, which would be much quicker after the initial clone.

Reply:
Aaah well.  Thanks for checking.  I suppose a possible future step is to detect if partial clones are supported by both git client and server, and use them if so...but caching could be the more helpful next step.